net/libfab/libfab.c

```diff
@@ -0,0 +1,3455 @@
+			libfab_bufq_process(ma);
+		}
+		break;
+	}
+
+	default:
+		M0_IMPOSSIBLE("invalid queue type: %x", nb->nb_qtype);
+		break;
+	}
+
+	if (ret == FI_SUCCESS) {
+		fbp->fb_state = FAB_BUF_QUEUED;
+		m0_tlink_init(&fab_bufhash_tl, fbp);
+		fab_bufhash_htable_add(&ma->ftm_bufhash.bht_hash, fbp);
+	}
+
+	return M0_RC(ret);
+}
+
+/**
+ * Cancels a buffer operation.
+ *
+ * Used as m0_net_xprt_ops::xo_buf_del().
+ *
+ * @see m0_net_buffer_del().
+ */
+static void libfab_buf_del(struct m0_net_buffer *nb)
+{
+	struct m0_fab__buf *buf = nb->nb_xprt_private;
+	struct m0_fab__tm  *ma = libfab_buf_ma(nb);
+
+	M0_PRE(libfab_tm_is_locked(ma) && libfab_tm_invariant(ma) &&
+	       libfab_buf_invariant(buf));
+	nb->nb_flags |= M0_NET_BUF_CANCELLED;
+
+	if (buf->fb_state != FAB_BUF_QUEUED)
+		buf->fb_state = FAB_BUF_QUEUED;
+
+	libfab_buf_done(buf, -ECANCELED);
+}
+
+/**
+ * Used as m0_net_xprt_ops::xo_bev_deliver_sync().
+ *
+ * @see m0_net_bev_deliver_sync().
+ */
+static int libfab_bev_deliver_sync(struct m0_net_transfer_mc *ma)
+{
+	return 0;
+}
+
+/**
+ * Used as m0_net_xprt_ops::xo_bev_deliver_all().
+ *
+ * @see m0_net_bev_deliver_all().
+ */
+static void libfab_bev_deliver_all(struct m0_net_transfer_mc *ma)
+{
+
+}
+
+/**
+ * Used as m0_net_xprt_ops::xo_bev_pending().
+ *
+ * @see m0_net_bev_pending().
+ */
+static bool libfab_bev_pending(struct m0_net_transfer_mc *ma)
+{
+	return false;
+}
+
+/**
+ * Used as m0_net_xprt_ops::xo_bev_notify().
+ *
+ * @see m0_net_bev_notify().
+ */
+static void libfab_bev_notify(struct m0_net_transfer_mc *ma,
+			      struct m0_chan *chan)
+{
+
+}
+
+/**
+ * Maximal number of bytes in a buffer.
+ *
+ * Used as m0_net_xprt_ops::xo_get_max_buffer_size()
+ *
+ * @see m0_net_domain_get_max_buffer_size()
+ */
+static m0_bcount_t libfab_get_max_buf_size(const struct m0_net_domain *dom)
+{
+	struct m0_fab__ndom *nd = dom->nd_xprt_private;
+
+	return (m0_bcount_t)(nd->fnd_seg_size * nd->fnd_seg_nr);
+}
+
+/**
+ * Maximal number of bytes in a buffer segment.
+ *
+ * Used as m0_net_xprt_ops::xo_get_max_buffer_segment_size()
+ *
+ * @see m0_net_domain_get_max_buffer_segment_size()
+ */
+static m0_bcount_t libfab_get_max_buf_seg_size(const struct m0_net_domain *dom)
+{
+	return ((struct m0_fab__ndom *)dom->nd_xprt_private)->fnd_seg_size;
+}
+
+/**
+ * Maximal number of segments in a buffer
+ *
+ * Used as m0_net_xprt_ops::xo_get_max_buffer_segments()
+ *
+ * @see m0_net_domain_get_max_buffer_segments()
+ */
+static int32_t libfab_get_max_buf_segments(const struct m0_net_domain *dom)
+{
+	return ((struct m0_fab__ndom *)dom->nd_xprt_private)->fnd_seg_nr;
+}
+/**
+ * Maximal number of bytes in a buffer descriptor.
+ *
+ * Used as m0_net_xprt_ops::xo_get_max_buffer_desc_size()
+ *
+ * @see m0_net_domain_get_max_buffer_desc_size()
+ */
+static m0_bcount_t libfab_get_max_buf_desc_size(const struct m0_net_domain *dom)
+{
+	struct m0_fab__ndom *nd = dom->nd_xprt_private;
+
+	return (m0_bcount_t)(sizeof(struct m0_fab__bdesc) + 
```

Comment:
Cast is useless. Overflow can happen at multiplication if size_t is 32 bits.

Reply:
Corrected